### PR TITLE
Move outer joins to JoinLoop framework

### PR DESCRIPTION
**Pull request by asuhan**

### CODE REVIEW

File: QueryEngine/IRCodegen.cpp [hunk 1]
```diff
@@ -353,26 +339,100 @@ std::vector<JoinLoop> Executor::buildJoinLoops(RelAlgExecutionUnit& ra_exe_unit,
...
-        const auto rows_per_scan_ptr = cgen_state_->ir_builder_.CreateGEP(
-            get_arg_by_name(cgen_state_->row_func_, "num_rows_per_scan"), ll_int(int32_t(level_idx + 1)));
-        domain.upper_bound = cgen_state_->ir_builder_.CreateLoad(rows_per_scan_ptr, "num_rows_per_scan");
-        return domain;
-      });
+      // Callback provided to the `JoinLoop` framework to evaluate the (outer) join condition.
+      const auto outer_join_condition_cb =
+          [this, level_idx, &co, &current_level_join_conditions](const std::vector<llvm::Value*>& prev_iters) {
+            // The values generated for the match path don't dominate all uses
+            // since on the non-match path nulls are generated. Reset the cache
```
[the remark this answers is not in the document]
Comment (asuhan) [reply QueryEngine/IRCodegen.cpp]: They're generated by the `ColumnVar` codegen, more precisely `Executor::codegenOuterJoinNullPlaceholder`. The `JoinLoop` framework is only responsible for letting the users know if there was a match or not, the way nulls are generated is outside its control.

File: QueryEngine/IRCodegen.cpp [hunk 2]
```diff
@@ -353,26 +339,100 @@ std::vector<JoinLoop> Executor::buildJoinLoops(RelAlgExecutionUnit& ra_exe_unit,
...
+          current_level_join_conditions.type,
+          [this, level_idx, &co](const std::vector<llvm::Value*>& prev_iters) {
+            addJoinLoopIterator(prev_iters, level_idx);
+            JoinLoopDomain domain{0};
+            const auto rows_per_scan_ptr = cgen_state_->ir_builder_.CreateGEP(
+                get_arg_by_name(cgen_state_->row_func_, "num_rows_per_scan"), ll_int(int32_t(level_idx + 1)));
+            domain.upper_bound = cgen_state_->ir_builder_.CreateLoad(rows_per_scan_ptr, "num_rows_per_scan");
+            return domain;
+          },
+          current_level_join_conditions.type == JoinType::LEFT
```
[the remark this answers is not in the document]
Comment (asuhan) [reply QueryEngine/IRCodegen.cpp]: Correct, it's just `INNER` and `LEFT` (have a look at the enum).

File: QueryEngine/LoopControlFlow/JoinLoopTest.cpp [hunk 3]
```diff
@@ -147,18 +148,23 @@ std::vector<JoinLoop> generate_descriptors(const unsigned mask,
...
                                 return domain;
                               },
+                              nullptr,
+                              nullptr,
                               "i" + std::to_string(i));
       ++cond_idx;
     } else {
       const auto upper_bound = upper_bounds[i];
       join_loops.emplace_back(JoinLoopKind::UpperBound,
+                              JoinType::INNER,
```
[the remark this answers is not in the document]
Comment (asuhan) [reply QueryEngine/LoopControlFlow/JoinLoopTest.cpp]: Yes, I'm still trying to find the best way to test this cleanly with not much success yet. A lot of the stuff which could go wrong with the outer joins is in the callbacks, which makes it a challenge to test it meaningfully in isolation. Pretty sure it'll have to be a separate PR, I expect the setup for a good test to be quite onerous.

File: QueryEngine/ColumnIR.cpp [hunk 4]
```diff
@@ -83,7 +83,9 @@ int adjusted_range_table_index(const Analyzer::ColumnVar* col_var) {
 std::vector<llvm::Value*> Executor::codegen(const Analyzer::ColumnVar* col_var,
                                             const bool fetch_column,
                                             const CompilationOptions& co) {
-  if (!cgen_state_->outer_join_cond_lv_ || col_var->get_rte_idx() == 0) {
+  if (col_var->get_rte_idx() <= 0 ||
+      (!cgen_state_->outer_join_cond_lv_ &&
+       (cgen_state_->outer_join_match_found_per_level_.empty() || !foundOuterJoinMatch(col_var->get_rte_idx())))) {
```
[the remark this answers is not in the document]
Comment (asuhan) [reply QueryEngine/ColumnIR.cpp]: Yes, that's almost the direct definition of outer joins (or at least my understanding of it) -- columns which don't participate in an outer joins have unchanged behavior.

**Comment by shtilman** [comment]
ok, @andrewseidl, changes look good to me.
LEFT tests will be added separately

File: QueryEngine/RelAlgTranslator.cpp [hunk 5]
```diff
@@ -299,7 +299,7 @@ std::shared_ptr<Analyzer::Expr> RelAlgTranslator::translateInput(const RexInput*
       CHECK_EQ("rowid", cd->columnName);
       col_ti.set_size(8);
     }
-    if (rte_idx > 0 && join_type_ == JoinType::LEFT) {
+    if (rte_idx > 0 && join_types_[rte_idx - 1] == JoinType::LEFT) {
```
Comment (andrewseidl) [inline QueryEngine/RelAlgTranslator.cpp]: ASan seems to be failing here (and I'd assume at line 312 below).
[asan.txt](https://github.com/mapd/mapd-core/files/1644899/asan.txt)

Comment (asuhan) [reply QueryEngine/RelAlgTranslator.cpp]: Thanks, I'm looking into it right now.

**Comment by asuhan** [comment]
@andrewseidl https://github.com/mapd/mapd-core/pull/175/commits/7db83d3e414726ce403d3e07590cb74680fa0167 fixes it.

**Comment by andrewseidl** [comment]
Squashed the asan fix into `3/5`, the commit which introduced the issue.

Thanks!

**Pull request merged**